beacon_node/http_api/src/lib.rs

```diff
@@ -402,40 +403,71 @@ pub fn serve<T: BeaconChainTypes>(
-                            .map(|(index, (validator, balance))| api_types::ValidatorData {
-                                index: index as u64,
-                                balance: *balance,
-                                status: api_types::ValidatorStatus::from_validator(
-                                    Some(validator),
-                                    epoch,
-                                    finalized_epoch,
-                                    far_future_epoch,
-                                ),
-                                validator: validator.clone(),
-                            })
-                            .collect::<Vec<_>>())
-                    })
-                    .map(api_types::GenericResponse::from)
-            })
-        });
+        .and_then(
+            |state_id: StateId, chain: Arc<BeaconChain<T>>, query: api_types::ValidatorsQuery| {
+                blocking_json_task(move || {
+                    let id_predicate = |index: &u64, validator: &Validator| match query.id.as_ref()
+                    {
+                        Some(ids) => ids.0.iter().any(|id| match &id {
+                            ValidatorId::PublicKey(param_pubkey) => {
+                                validator.pubkey == *param_pubkey
+                            }
+                            ValidatorId::Index(param_index) => index == param_index,
+                        }),
+                        None => true,
+                    };
+                    let status_predicate =
+                        |validator_status: &ValidatorStatus| match query.status.as_ref() {
+                            Some(statuses) => statuses
+                                .0
+                                .iter()
+                                .any(|param_status| param_status == validator_status),
+                            None => true,
+                        };
+                    state_id
+                        .map_state(&chain, |state| {
+                            let epoch = state.current_epoch();
+                            let finalized_epoch = state.finalized_checkpoint.epoch;
+                            let far_future_epoch = chain.spec.far_future_epoch;
+
+                            Ok(state
+                                .validators
+                                .iter()
+                                .zip(state.balances.iter())
+                                .enumerate()
+                                .filter_map(|(index, (validator, balance))| {
+                                    let status = api_types::ValidatorStatus::from_validator(
+                                        Some(validator),
+                                        epoch,
+                                        finalized_epoch,
+                                        far_future_epoch,
+                                    );
+
+                                    if id_predicate(&(index as u64), validator)
```

Comment:
I would place these predicates (perhaps as `map_or`s) above the `filter_map` on L444; this will stop us from allocating a `ValidatorStatus` for validators we're not going to return.

Reply:
I moved the filters ahead of the map, but because the second filter tests against `ValidatorStatus`, am I now allocating a `ValidatorStatus` twice whenever we return a validator that passes a status filter ?